### ISSUE TITLE
Incorrect condition behavior for tasks making use of failed_when statement

### ISSUE THREAD

**Issue by timmjd** [ansible/ansible#28455]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
- Conditional & Error Handling

##### ANSIBLE VERSION
```
ansible 2.3.1.0
  configured module search path = Default w/o overrides
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
```

##### CONFIGURATION
N/A

##### OS / ENVIRONMENT
N/A

##### SUMMARY
Making use of the [failed_when](http://docs.ansible.com/ansible/latest/playbooks_error_handling.html#handlers-and-failure) does only affect the state of the current task. Verifying the result within a [condition](http://docs.ansible.com/ansible/latest/playbooks_conditionals.html) of another task does ignore the `failed_when` modification.

##### STEPS TO REPRODUCE
Verifying a SSH connection connections to github will return 1 on success and therefore officially fail. Modifying the behavior via `failed_when` to check the if SSH login was successful does work for the task itself (will only fail if credentials are not set). Any other task that is reading the result within the `register` will get the result of the the unmodified behavior.
* **Expected** behavior: `(result | succeeded) == true`
* **Actual** behavior: `(result | failed) == true`
* **Workaround** `(result.failed) != true`

```yaml
- name: Check SSH connection
  command: 'ssh git@github.com'
  register: result
  ignore_errors: true
  failed_when: (result.stderr | search('Permission denied'))
- debug:
    msg: '{{result}}'
- debug:
    msg: 'Expected'
  when: result | succeeded
- debug:
    msg: 'Actual'
  when: result | failed
- debug:
    msg: 'Workarround'
  when: not result.failed
```

Output
```
TASK [Check SSH connection] ****************************************************
changed: [localhost]

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": {
        "changed": true,
        "cmd": [
            "ssh",
            "git@github.com"
        ],
        "delta": "0:00:00.543289",
        "end": "2017-08-20 10:11:53.085470",
        "failed": false,
        "failed_when_result": false,
        "rc": 1,
        "start": "2017-08-20 10:11:52.542181",
        "stderr": "Pseudo-terminal will not be allocated because stdin is not a terminal.\r\nHi xxx! You've successfully authenticated, but GitHub does not provide shell access.",
        "stderr_lines": [
            "Pseudo-terminal will not be allocated because stdin is not a terminal.",
            "Hi xxx! You've successfully authenticated, but GitHub does not provide shell access."
        ],
        "stdout": "",
        "stdout_lines": []
    }
}

TASK [debug] *******************************************************************
skipping: [localhost]

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": "Actual"
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": "Workarround"
}
```

This behavior differs from other examples where `result` is consistent for later functions
```yaml
- name: Check Whatever
  stat:
    path: /
  register: result
- debug:
    msg: '{{result}}'
- debug:
    msg: 'succeeded'
  when: result | succeeded
- debug:
    msg: 'failed'
  when: result | failed
```

Output
```
TASK [Check Whatever] **********************************************************
ok: [localhost]

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": {
        "changed": false,
        "stat": {
            ...
        }
    }
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": "succeeded"
}

TASK [debug] *******************************************************************
skipping: [localhost]
```

##### EXPECTED RESULTS
The state of a task that is forwarded by `register` can be verified with `succeeded` or `failed` outside the current task. A condition checking the result should get the same result as the actual task.

##### ACTUAL RESULTS
Only the current task respects the `failed_when` option. Other tasks do the result like if the task would have failed.

**Comment by nrwahl2** [comment]
Fixed in #26445 for version 2.4.

As you've probably noticed, the tests are checking different criteria in version 2.3. `results | failed` is checking the return code, while `results.failed` is checking the `failed` key directly.

The relevant file is `ansible/lib/ansible/plugins/test/core.py`.

Version 2.3:
```
def failed(*a, **kw):
    ''' Test if task result yields failed '''
    item = a[0]
    if type(item) != dict:
        raise errors.AnsibleFilterError("|failed expects a dictionary")
    rc = item.get('rc',0)
    failed = item.get('failed',False)
    if rc != 0 or failed:
        return True
    else:
        return False
```

Version 2.4, devel:
```
def failed(result):
    ''' Test if task result yields failed '''
    if not isinstance(result, MutableMapping):
        raise errors.AnsibleFilterError("|failed expects a dictionary")
    return result.get('failed', False)
```


This is the behavior when running from 2.4 devel:
```
PLAY [localhost] *************************************************************************************************************

TASK [Check SSH connection] **************************************************************************************************
changed: [localhost]

TASK [debug] *****************************************************************************************************************
ok: [localhost] => {
    "msg": {
        "changed": true,
        "cmd": [
            "ssh",
            "git@github.com"
        ],
        "delta": "0:00:00.062634",
        "end": "2017-08-20 18:04:09.268522",
        "failed": false,
        "failed_when_result": false,
        "msg": "non-zero return code",
        "rc": 255,
        "start": "2017-08-20 18:04:09.205888",
        "stderr": "Pseudo-terminal will not be allocated because stdin is not a terminal.\r\nssh: Could not resolve hostname github.com: Name or service not known",
        "stderr_lines": [
            "Pseudo-terminal will not be allocated because stdin is not a terminal.",
            "ssh: Could not resolve hostname github.com: Name or service not known"
        ],
        "stdout": "",
        "stdout_lines": []
    }
}

TASK [debug] *****************************************************************************************************************
ok: [localhost] => {
    "msg": "Expected"
}

TASK [debug] *****************************************************************************************************************
skipping: [localhost]

TASK [debug] *****************************************************************************************************************
ok: [localhost] => {
    "msg": "Workarround"
}

PLAY RECAP *******************************************************************************************************************
localhost                  : ok=4    changed=1    unreachable=0    failed=0
```

**Comment by nrwahl2** [comment]
It should be noted that there was significant internal debate in the PR that I referenced, so the future of this behavior is unknown.

**Comment by bcoca** [comment]
@nrwahl2 this is a different issue, the `failed_when` affects task status, but does not update the registered var.

**Comment by sivel** [comment]
As noted above, this functionality has been fixed in recent versions of Ansible (2.4+). If you are still experiencing problems with a recent version of ansible, please feel free to open a new issue.

If you have further questions please stop by IRC or the mailing list:

   * IRC: #ansible on irc.freenode.net
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project
